### PR TITLE
base-images: Migrate Debian base images to k/release

### PR DESCRIPTION
**Pull request by justaugustus**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Similar to https://github.com/kubernetes/release/pull/1140, I don't believe we need to/should be building the Debian base images from kubernetes/kubernetes.

I [submitted a PR](https://github.com/kubernetes/kubernetes/pull/93562) around 20 hours ago to rebuild the debian-base image to patch a CVE (ref: https://github.com/kubernetes/kubernetes/issues/92616, https://github.com/kubernetes/kubernetes/issues/93114) and it still hadn't merged (at the time I originally wrote up https://github.com/kubernetes/release/pull/1450).

While that is a fix for a medium CVE, this is not a great place to be in from a security standpoint.

This needs to be merged alongside:
- https://github.com/kubernetes/release/pull/1450
- https://github.com/kubernetes/test-infra/pull/18558

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

This used to be a PR to stop build base images altogether, but now it's just a move to k/release.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: build/debian-iptables/iptables-wrapper [hunk 1]
```diff
@@ -1,44 +0,0 @@
-#!/bin/sh
```
Comment (BenTheElder) [inline build/debian-iptables/iptables-wrapper]: we need this

Comment (BenTheElder) [reply build/debian-iptables/iptables-wrapper]: (also we need both legacy and non legacy iptables binaries)

**Comment by BenTheElder** [comment]
we can't totally get away from building our own FWIW, due to https://github.com/kubernetes/kubernetes/issues/71305

we can't use a boring plain ol image that contains iptables, we have to employ @danwinship's hacks :+)

bonus: https://github.com/kubernetes/kubernetes/issues/82587

**Comment by danwinship** [comment]
The iptables hack could be moved into the kube-proxy image rather than being in a base image. I don't think any other (official) images need the hack.

**Comment by justaugustus** [comment]
BWUUAHHAHA, these are the kind of useful tidbits I needed. Thanks y'all!

**Comment by BenTheElder** [comment]
> The iptables hack could be moved into the kube-proxy image rather than being in a base image. I don't think any other (official) images need the hack.

that's fair, though some less official ones probably do:
- ip-masq-agent (owned by sig-network?)
- kindnetd (don't consider this too much, we can do this ourselves, BUT it is using it)
- possibly others ...?

**Comment by justaugustus** [comment]
We're not going to land this for 1.18.
/milestone v1.19

**Comment by BenTheElder** [comment]
we don't build any of the debian images in normal CI so that seems irrelevant now

**Comment by BenTheElder** [comment]
er, but it is desirable to keep the "binary in a container" image builds to just `FROM BASE; COPY BINARY`, we should actually be able to do that without docker these days and it's much cheaper / faster, simpler to cross build ...

**Comment by danwinship** [comment]
Tim's comment about not failing on apt-get in test runs means we may need to keep the debian-iptables image, but FTR, the iptables hack is now available from https://github.com/kubernetes-sigs/iptables-wrappers/ for use by third-party packages, so we don't necessarily need to worry about other people needing debian-iptables.

**Comment by tallclair** [comment]
I'd like to see this move forward in v1.19 (I actually want to discuss cherry-picks, but I'll leave that for a different discussion).

I propose we revert the debian-iptables removal here, and open a separate PR or issue to figure out what to do with debian-iptables. For now, we could just build debian-iptables on the same base we choose for everything else (I think debian:buster-slim?)

Aside from debian-iptables, what is left for this PR to drop the [WIP] tag? We still need to modify a bunch of dockerfiles to change from debian-base to debian:buster-slim, were you planning to add that to this PR or open a separate one?

**Comment by justaugustus** [comment]
@tallclair -- I'm coming back to this soon; just handling some start of cycle administrivia which has priority at the moment.

**Comment by justaugustus** [comment]
Explicit hold:
/hold

@tallclair --

> I'd like to see this move forward in v1.19 (I actually want to discuss cherry-picks, but I'll leave that for a different discussion).
>
> I propose we revert the debian-iptables removal here, and open a separate PR or issue to figure out what to do with debian-iptables. For now, we could just build debian-iptables on the same base we choose for everything else (I think debian:buster-slim?)
> 
> Aside from debian-iptables, what is left for this PR to drop the [WIP] tag? We still need to modify a bunch of dockerfiles to change from debian-base to debian:buster-slim, were you planning to add that to this PR or open a separate one?

I've updated this PR to only remove the `debian-base` image.
We've got tracking issues for both routes:

- distroless: https://github.com/kubernetes/kubernetes/issues/70249
- updating Debian base images: https://github.com/kubernetes/kubernetes/issues/90698

To remove the WIP, a few things need to happen:

- `go-runner` needs to replace `debian-base`: https://github.com/kubernetes/kubernetes/pull/90674
- Successful releases with new `go-runner` base image
- Replacement base image for `debian-iptables`

I think we need to see how @dims' PR plays out before doing anything else, since we need to ensure release continuity on the off-chance that we need to back out any `go-runner` changes.

Let's put this on the agenda for next week's Release Engineering meeting and see what the best course of action is.
cc: @kubernetes/release-engineering

**Comment by kbruner** [comment]
Hello from the bug triage team! This issue has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting June 25th (about 4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.19, is it still planned for this release?

**Comment by kbruner** [comment]
Is this PR still targeted for 1.19 @justaugustus ?

**Comment by justaugustus** [comment]
As long as hyperkube is supported for older branches, we need to keep the base images around.
Hyperkube was deprecated in 1.18, its images are removed from 1.19 and on, which means we need to support the base images until 1.21.

/milestone v1.21

**Comment by justaugustus** [comment]
Slightly shifting the focus of this PR...

While we need to continue maintaining the Debian base images (primarily for earlier release branches), we don't need to maintain them in kubernetes/kubernetes.

Patching anything related to these images is expensive, given the time we wait for unnecessary presubmits to pass.

I'm proposing that we move maintenance of these images to k/release, as we did with kube-cross in https://github.com/kubernetes/release/pull/1140 / https://github.com/kubernetes/kubernetes/pull/89275.

I've opened https://github.com/kubernetes/release/pull/1450 and https://github.com/kubernetes/test-infra/pull/18558 to do this.

/priority critical-urgent
/milestone v1.19

**Comment by justaugustus** [comment]
/assign @dims @tallclair 
cc: @kubernetes/release-engineering 
ref: https://github.com/kubernetes/kubernetes/pull/88603#issuecomment-666980658

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Review by cpanato — APPROVED** [review]
/lgtm

**Review by xmudrii — APPROVED** [review]
/lgtm

**Review by hasheddan — APPROVED** [review]
/lgtm

**Comment by BenTheElder** [comment]
/retitle [WIP] base-images: Migrate Debian base images to k/release
> Replacement base image for debian-iptables

I don't see this happening easily with removing debian-base.

**Comment by liggitt** [comment]
does this need to be in 1.19? there seem to be enough actual release-blocking items in flight that it would be great to defer optional things to 1.20

**Comment by justaugustus** [comment]
> does this need to be in 1.19? there seem to be enough actual release-blocking items in flight that it would be great to defer optional things to 1.20

I'd prefer to not have these linger around in another release branch, if it can be helped.
Given the test results, this code removal should have no impact to k/k.

**Comment by justaugustus** [comment]
/hold cancel
https://github.com/kubernetes/release/pull/1450 is merging.

**Pull request merged**